lambda/translcore.ml

```diff
@@ -467,247 +471,73 @@ and transl_exp0 ~in_new_scope ~scopes e =
-    let translated_blocks = 
-      List.fold_left transl_block base_block 
-        (List.map (fun a -> a.clauses) (List.tl blocks))
-    in
-    (match global_counter with
-        | None -> translated_blocks
-        | Some gc -> 
-          Llet(
-            Variable, Pintval, gc, int 0,
-            Llet(
-              Strict, Pgenval, arr,
-              translated_blocks,
-              transl_concat_arrays (Lvar(arr)) block_count gc
-            )))
+    let loc = (of_location ~scopes e.exp_loc) in
+    let array_kind = Typeopt.array_kind e in
+    transl_arr_comprehension body blocks ~array_kind ~scopes ~loc
   | Texp_list_comprehension (body, blocks) -> 
+    let loc = of_location ~scopes e.exp_loc in
     let int n = Lconst (Const_base (Const_int n)) in  
-    let block = List.fold_left (fun acc block -> 
-        let rec loop = function 
-          | [] -> assert false 
-          | [last] -> [last, block.guard]
-          | l::ls -> (l, None) :: (loop ls) 
-        in
-        List.concat [List.rev (loop block.clauses); acc]
-        ) [] blocks
-    in
-    let rec transl_list_comp block upper_res= 
-      let res = Ident.create_local "res" in
-      let is_concat, (type_comp, guard), bdy, return =
-        match block with
-        | [last] ->  
-          (* Cons element to the result list.*)
-          let bdy = 
-            Lprim( 
-              (*TODO: change None to usefull P...val*)
-              Pmakeblock(0, Immutable, None), 
-              [(transl_exp ~scopes  body); Lvar(res)], Loc_unknown) 
-          in
-          false, last, bdy, Typeopt.value_kind body.exp_env body.exp_type
-        | el::ls ->  
-          true, el, (transl_list_comp ls (Some (Lvar(res)))), Pgenval
-        | _ -> assert false
-      in
-      let bdy =   
-        match guard with 
-        | None -> bdy
-        | Some guard -> Lifthenelse((transl_exp ~scopes  guard), bdy, Lvar(res))  
-      in
-      (*Single loop optimization makes sure we dont create a list that has
-        to be reversed. We directly create the correct list.*)
-      (*There is no upper res if there is only a single loop.*)
-      let args =  match upper_res with
-        | None -> [] 
-        | Some ur -> [ur] 
-      in
-      let single_loop = Option.is_none upper_res in
-      let param, pval, args, func =
+    let transl_list_comp type_comp body acc_var = 
+      let new_acc = Ident.create_local "acc" in
+      let param, pval, args, func, body =
         match type_comp with
         | From_to (param, _,e2,e3, dir) -> 
             param, Pintval, 
-              (transl_exp ~scopes e2)::(transl_exp ~scopes e3)::args,
-              from_to_comp_prim ~is_concat ~single_loop ~dir
-        | In (param, _, e2) ->  
-            param, Pgenval, (transl_exp ~scopes e2)::args,
-              in_from_to_comp_prim ~is_concat ~single_loop
+            [transl_exp ~scopes e2; transl_exp ~scopes e3; Lvar(new_acc)],
+            from_to_comp_prim ~dir,
+            body
```

Comment:
These kinds of tuple expressions can be hard to read. I think it is generally better to write them as e.g.:
```ocaml
let pval = Pintval in
let args = [transl_exp ~scopes e2; transl_exp ~scopes e3; Lvar(new_acc)] in
let func = from_to_comp_prim ~dir in
param, pval, args, func, body
```
that way the components get meaningful names that match what they will be bound to be the surrounding let.